### PR TITLE
Add Q monoid instance to fix a compile failure

### PR DESCRIPTION
**Pull request by TristanCacqueray**

### CODE REVIEW

**Comment by TristanCacqueray** [comment]
I don't fully understand what is this all about, but I needed this change to make retroclash-lib compile with ghc-8.10.7 and template-haskell-2.16.0.0 .

**Comment by gergoerdi** [comment]
Hmm, it's weird because the Stack setup uses LTS 18.14 which [seems to contain `template-haskell 2.16.0.0`](https://www.stackage.org/lts-18.14/package/template-haskell-2.16.0.0).

**Comment by gergoerdi** [comment]
OK so building with Stack using the included `stack.yaml`, based on LTS 18.14, I get an error about duplicate instances:

```
retroclash-lib/src/RetroClash/Memory.hs:63:10: error:
    Duplicate instance declarations:
      instance Semigroup a => Semigroup (Q a)
        -- Defined at src/RetroClash/Memory.hs:63:10
      instance Semigroup a => Semigroup (Q a)
        -- Defined in ‘th-orphans-0.13.12:Language.Haskell.TH.Instances’
   |
63 | instance (Semigroup a) => Semigroup (Q a) where
   |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

retroclash-lib/src/RetroClash/Memory.hs:66:10: error:
    Duplicate instance declarations:
      instance Monoid a => Monoid (Q a)
        -- Defined at src/RetroClash/Memory.hs:66:10
      instance Monoid a => Monoid (Q a)
        -- Defined in ‘th-orphans-0.13.12:Language.Haskell.TH.Instances’
   |
66 | instance (Monoid a) => Monoid (Q a) where
   |          ^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by gergoerdi** [comment]
Can you try https://github.com/TristanCacqueray/retroclash-lib/commit/c62f0adfab5cf39e2de0133b1960385d607968ee?

**Comment by TristanCacqueray** [comment]
Thanks, c62f0ad works for me.

File: src/RetroClash/Memory.hs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-{-# LANGUAGE DerivingStrategies, GeneralizedNewtypeDeriving #-}
```
Comment (TristanCacqueray) [inline src/RetroClash/Memory.hs]: the CPP can be removed

**Comment by gergoerdi** [comment]
> Thanks, [c62f0ad](https://github.com/gergoerdi/retroclash-lib/commit/c62f0adfab5cf39e2de0133b1960385d607968ee) works for me.

OK, thanks. TBH I still don't understand what exactly is going on, since the Stack build uses Stackage LTS-18.14 which has GCH 8.10.7 and `template-haskell` 2.16.0.0, same versions that you're trying to use. Even more mysterious is that when I add your patch, I was getting an instance conflict from `th-orphans-0.13.12:Language.Haskell.TH.Instances` which I was not importing. So what is re-exporting its instances for me but not for you?

Anyway, since we now have a version that works for both of us, I'll merge that and close this.

File: src/RetroClash/Memory.hs (hunk 1 above)
Comment (gergoerdi) [reply src/RetroClash/Memory.hs]: Good catch.

**Comment by TristanCacqueray** [comment]
Could it be that `th-orphans` is transitively pulled in by stack? I think your fix is correct, adding the depends on seem to ensure the Monoid instance is available. Thank you for merging the change!

For what its worth, the issue can be reproduced with this nix command:

```
NIXPKGS_ALLOW_BROKEN=1 nix-shell -I nixpkgs=https://github.com/NixOS/nixpkgs/archive/d156db7d1752b48531b91d65db8f9bc91e29c399.tar.gz \
  -p "(haskell.packages.ghc8107.override { overrides = \
f: p: { clash-prelude = (haskell.lib.dontCheck p.clash-prelude); };})\
.ghcWithPackages(p: [p.retroclash-lib])"
```

… which produces this output:

```
building '/nix/store/xgakbif7d20j1rjar925r426pr4br37x-retroclash-lib-0.1.2.drv'...
setupCompilerEnvironmentPhase
Build with /nix/store/fzg5mzwjdf9rln0srih4zy55d1p04hry-ghc-8.10.7.
unpacking sources
unpacking source archive /nix/store/8id4vixyay4amcv8pihdbvl5vl9b9mz9-retroclash-lib-0.1.2.tar.gz
source root is retroclash-lib-0.1.2
setting SOURCE_DATE_EPOCH to timestamp 1650717177 of file retroclash-lib-0.1.2/retroclash-lib.cabal
patching sources
compileBuildDriverPhase
setupCompileFlags: -package-db=/tmp/nix-build-retroclash-lib-0.1.2.drv-0/setup-package.conf.d -j8 +RTS -A64M -RTS -threaded -rtsopts
[1 of 1] Compiling Main             ( /nix/store/4mdp8nhyfddh7bllbi7xszz7k9955n79-Setup.hs, /tmp/nix-build-retroclash-lib-0.1.2.drv-0/Main.o )
Linking Setup ...
configuring
configureFlags: --verbose --prefix=/nix/store/9vnmy398wqkkgq48yip0473409wf94n6-retroclash-lib-0.1.2 --libdir=$prefix/lib/$compiler --libsubdir=$abi/$libname --docdir=/nix/store/aaw1sjzmdrn28dibyf8hliiawp2w73ch-retroclash-lib-0.1.2-doc/share/doc/retroclash-lib-0.1.2 --with-gcc=gcc --package-db=/tmp/nix-build-retroclash-lib-0.1.2.drv-0/package.conf.d --ghc-options=-j8 +RTS -A64M -RTS --disable-split-objs --enable-library-profiling --profiling-detail=exported-functions --disable-profiling --enable-shared --disable-coverage --enable-static --disable-executable-dynamic --enable-tests --disable-benchmarks --enable-library-vanilla --disable-library-for-ghci --ghc-option=-split-sections --extra-lib-dirs=/nix/store/5xriv7q6sadqlsl5wkcnl39wzyimk8vy-ncurses-6.3/lib --extra-lib-dirs=/nix/store/ryv1fym9ms3lbniq4clicdc781wy9rap-libffi-3.4.2/lib --extra-lib-dirs=/nix/store/ddgmlyaz6adjvn7pkic3gr9xw94zs0l2-gmp-with-cxx-6.2.1/lib --extra-lib-dirs=/nix/store/fzg5mzwjdf9rln0srih4zy55d1p04hry-ghc-8.10.7/lib
Using Parsec parser
Configuring retroclash-lib-0.1.2...
Dependency barbies >=2.0.1 && <2.1: using barbies-2.0.3.1
Dependency base >=4.14 && <5: using base-4.14.3.0
Dependency clash-ghc >=1.4.2 && <1.5 || >=1.6.1 && <1.7: using clash-ghc-1.6.3
Dependency clash-lib >=1.4.2 && <1.5 || >=1.6.1 && <1.7: using clash-lib-1.6.3
Dependency clash-prelude >=1.4.2 && <1.5 || >=1.6.1 && <1.7: using
clash-prelude-1.6.3
Dependency containers -any: using containers-0.6.5.1
Dependency ghc-typelits-extra -any: using ghc-typelits-extra-0.4.3
Dependency ghc-typelits-knownnat -any: using ghc-typelits-knownnat-0.7.6
Dependency ghc-typelits-natnormalise -any: using
ghc-typelits-natnormalise-0.7.6
Dependency lens -any: using lens-5.0.1
Dependency lift-type ==0.1.*: using lift-type-0.1.0.1
Dependency monoidal-containers ==0.6.*: using monoidal-containers-0.6.2.0
Dependency mtl -any: using mtl-2.2.2
Dependency template-haskell -any: using template-haskell-2.16.0.0
Dependency transformers -any: using transformers-0.5.6.2
Source component graph: component lib
Configured component graph:
    component retroclash-lib-0.1.2-IoXGmPAf0NPGHF1IUmipkF
        include barbies-2.0.3.1-7ZlPOuzXrhD6ffhMHoqD5c
        include base-4.14.3.0
        include clash-ghc-1.6.3-43Rj1Jm638uGzifkOaGlUq
        include clash-lib-1.6.3-CCd61OZzP5VJqN42VzDsY7
        include clash-prelude-1.6.3-70ftuOdy50tI4TGDLYmM14
        include containers-0.6.5.1
        include ghc-typelits-extra-0.4.3-FiwzhcBQsP05qjdI5TTi7a
        include ghc-typelits-knownnat-0.7.6-5B9eT69FNqQGYwYAhU2BcD
        include ghc-typelits-natnormalise-0.7.6-4AFZKSzvdaDDKxkoU1eFO0
        include lens-5.0.1-5JHTxTLNLpj2b18jI2H5bb
        include lift-type-0.1.0.1-GZRqI1F776LFQ24koiUDLI
        include monoidal-containers-0.6.2.0-qNfDV2643RLdNU1Vrbzzj
        include mtl-2.2.2
        include template-haskell-2.16.0.0
        include transformers-0.5.6.2
Linked component graph:
    unit retroclash-lib-0.1.2-IoXGmPAf0NPGHF1IUmipkF
        include barbies-2.0.3.1-7ZlPOuzXrhD6ffhMHoqD5c
        include base-4.14.3.0
        include clash-ghc-1.6.3-43Rj1Jm638uGzifkOaGlUq
        include clash-lib-1.6.3-CCd61OZzP5VJqN42VzDsY7
        include clash-prelude-1.6.3-70ftuOdy50tI4TGDLYmM14
        include containers-0.6.5.1
        include ghc-typelits-extra-0.4.3-FiwzhcBQsP05qjdI5TTi7a
        include ghc-typelits-knownnat-0.7.6-5B9eT69FNqQGYwYAhU2BcD
        include ghc-typelits-natnormalise-0.7.6-4AFZKSzvdaDDKxkoU1eFO0
        include lens-5.0.1-5JHTxTLNLpj2b18jI2H5bb
        include lift-type-0.1.0.1-GZRqI1F776LFQ24koiUDLI
        include monoidal-containers-0.6.2.0-qNfDV2643RLdNU1Vrbzzj
        include mtl-2.2.2
        include template-haskell-2.16.0.0
        include transformers-0.5.6.2
        RetroClash.BCD=retroclash-lib-0.1.2-IoXGmPAf0NPGHF1IUmipkF:RetroClash.BCD,RetroClash.Barbies=retroclash-lib-0.1.2-IoXGmPAf0NPGHF1IUmipkF:RetroClash.Barbies,RetroClash.CPU=retroclash-lib-0.1.2-IoXGmPAf0NPGHF1IUmipkF:RetroClash.CPU,RetroClash.Clock=retroclash-lib-0.1.2-IoXGmPAf0NPGHF1IUmipkF:RetroClash.Clock,RetroClash.Delayed=retroclash-lib-0.1.2-IoXGmPAf0NPGHF1IUmipkF:RetroClash.Delayed,RetroClash.I2C=retroclash-lib-0.1.2-IoXGmPAf0NPGHF1IUmipkF:RetroClash.I2C,RetroClash.Keypad=retroclash-lib-0.1.2-IoXGmPAf0NPGHF1IUmipkF:RetroClash.Keypad,RetroClash.Memory=retroclash-lib-0.1.2-IoXGmPAf0NPGHF1IUmipkF:RetroClash.Memory,RetroClash.PS2=retroclash-lib-0.1.2-IoXGmPAf0NPGHF1IUmipkF:RetroClash.PS2,RetroClash.PS2.ASCII=retroclash-lib-0.1.2-IoXGmPAf0NPGHF1IUmipkF:RetroClash.PS2.ASCII,RetroClash.Port=retroclash-lib-0.1.2-IoXGmPAf0NPGHF1IUmipkF:RetroClash.Port,RetroClash.SerialRx=retroclash-lib-0.1.2-IoXGmPAf0NPGHF1IUmipkF:RetroClash.SerialRx,RetroClash.SerialTx=retroclash-lib-0.1.2-IoXGmPAf0NPGHF1IUmipkF:RetroClash.SerialTx,RetroClash.SevenSegment=retroclash-lib-0.1.2-IoXGmPAf0NPGHF1IUmipkF:RetroClash.SevenSegment,RetroClash.Stack=retroclash-lib-0.1.2-IoXGmPAf0NPGHF1IUmipkF:RetroClash.Stack,RetroClash.Utils=retroclash-lib-0.1.2-IoXGmPAf0NPGHF1IUmipkF:RetroClash.Utils,RetroClash.VGA=retroclash-lib-0.1.2-IoXGmPAf0NPGHF1IUmipkF:RetroClash.VGA,RetroClash.Video=retroclash-lib-0.1.2-IoXGmPAf0NPGHF1IUmipkF:RetroClash.Video
Ready component graph:
    definite retroclash-lib-0.1.2-IoXGmPAf0NPGHF1IUmipkF
        depends barbies-2.0.3.1-7ZlPOuzXrhD6ffhMHoqD5c
        depends base-4.14.3.0
        depends clash-ghc-1.6.3-43Rj1Jm638uGzifkOaGlUq
        depends clash-lib-1.6.3-CCd61OZzP5VJqN42VzDsY7
        depends clash-prelude-1.6.3-70ftuOdy50tI4TGDLYmM14
        depends containers-0.6.5.1
        depends ghc-typelits-extra-0.4.3-FiwzhcBQsP05qjdI5TTi7a
        depends ghc-typelits-knownnat-0.7.6-5B9eT69FNqQGYwYAhU2BcD
        depends ghc-typelits-natnormalise-0.7.6-4AFZKSzvdaDDKxkoU1eFO0
        depends lens-5.0.1-5JHTxTLNLpj2b18jI2H5bb
        depends lift-type-0.1.0.1-GZRqI1F776LFQ24koiUDLI
        depends monoidal-containers-0.6.2.0-qNfDV2643RLdNU1Vrbzzj
        depends mtl-2.2.2
        depends template-haskell-2.16.0.0
        depends transformers-0.5.6.2
Using Cabal-3.2.1.0 compiled by ghc-8.10
Using compiler: ghc-8.10.7
Using install prefix:
/nix/store/9vnmy398wqkkgq48yip0473409wf94n6-retroclash-lib-0.1.2
Executables installed in:
/nix/store/9vnmy398wqkkgq48yip0473409wf94n6-retroclash-lib-0.1.2/bin
Libraries installed in:
/nix/store/9vnmy398wqkkgq48yip0473409wf94n6-retroclash-lib-0.1.2/lib/ghc-8.10.7/x86_64-linux-ghc-8.10.7/retroclash-lib-0.1.2-IoXGmPAf0NPGHF1IUmipkF
Dynamic Libraries installed in:
/nix/store/9vnmy398wqkkgq48yip0473409wf94n6-retroclash-lib-0.1.2/lib/ghc-8.10.7/x86_64-linux-ghc-8.10.7
Private executables installed in:
/nix/store/9vnmy398wqkkgq48yip0473409wf94n6-retroclash-lib-0.1.2/libexec/x86_64-linux-ghc-8.10.7/retroclash-lib-0.1.2
Data files installed in:
/nix/store/9vnmy398wqkkgq48yip0473409wf94n6-retroclash-lib-0.1.2/share/x86_64-linux-ghc-8.10.7/retroclash-lib-0.1.2
Documentation installed in:
/nix/store/aaw1sjzmdrn28dibyf8hliiawp2w73ch-retroclash-lib-0.1.2-doc/share/doc/retroclash-lib-0.1.2
Configuration files installed in:
/nix/store/9vnmy398wqkkgq48yip0473409wf94n6-retroclash-lib-0.1.2/etc
No alex found
Using ar found on system at:
/nix/store/3fqi6nigj8dkbvjnw8y4dy59gkq8vsj4-binutils-2.38/bin/ar
No c2hs found
No cpphs found
No doctest found
Using gcc version 11.3.0 given by user at:
/nix/store/gfr4ljdlr1wplc01j4fchw4w9x3lfvv9-gcc-wrapper-11.3.0/bin/gcc
Using ghc version 8.10.7 found on system at:
/nix/store/fzg5mzwjdf9rln0srih4zy55d1p04hry-ghc-8.10.7/bin/ghc
Using ghc-pkg version 8.10.7 found on system at:
/nix/store/fzg5mzwjdf9rln0srih4zy55d1p04hry-ghc-8.10.7/bin/ghc-pkg
No ghcjs found
No ghcjs-pkg found
No greencard found
Using haddock version 2.24.2 found on system at:
/nix/store/fzg5mzwjdf9rln0srih4zy55d1p04hry-ghc-8.10.7/bin/haddock
No happy found
Using haskell-suite found on system at: haskell-suite-dummy-location
Using haskell-suite-pkg found on system at: haskell-suite-pkg-dummy-location
No hmake found
Using hpc version 0.68 found on system at:
/nix/store/fzg5mzwjdf9rln0srih4zy55d1p04hry-ghc-8.10.7/bin/hpc
Using hsc2hs version 0.68.7 found on system at:
/nix/store/fzg5mzwjdf9rln0srih4zy55d1p04hry-ghc-8.10.7/bin/hsc2hs
Using hscolour version 1.24 found on system at:
/nix/store/92f92g1w8yrf1ang83jjbsb8gmlr0hqn-hscolour-1.24.4/bin/HsColour
No jhc found
Using ld found on system at:
/nix/store/nhv0j3rsa1qirwgw9mym03lbbxy1almv-binutils-wrapper-2.38/bin/ld.gold
No pkg-config found
Using runghc version 8.10.7 found on system at:
/nix/store/fzg5mzwjdf9rln0srih4zy55d1p04hry-ghc-8.10.7/bin/runghc
Using strip version 2.38 found on system at:
/nix/store/gfr4ljdlr1wplc01j4fchw4w9x3lfvv9-gcc-wrapper-11.3.0/bin/strip
Using tar found on system at:
/nix/store/9ia80kzgnizdj7ninwqgswqw4fjk3kb9-gnutar-1.34/bin/tar
No uhc found
building
Preprocessing library for retroclash-lib-0.1.2..
Building library for retroclash-lib-0.1.2..
[ 1 of 19] Compiling Paths_retroclash_lib ( dist/build/autogen/Paths_retroclash_lib.hs, dist/build/Paths_retroclash_lib.o, dist/build/Paths_retroclash_lib.dyn_o )
[ 2 of 19] Compiling RetroClash.Barbies ( src/RetroClash/Barbies.hs, dist/build/RetroClash/Barbies.o, dist/build/RetroClash/Barbies.dyn_o )
[ 3 of 19] Compiling RetroClash.Clock ( src/RetroClash/Clock.hs, dist/build/RetroClash/Clock.o, dist/build/RetroClash/Clock.dyn_o )
[ 4 of 19] Compiling RetroClash.Port  ( src/RetroClash/Port.hs, dist/build/RetroClash/Port.o, dist/build/RetroClash/Port.dyn_o )
[ 5 of 19] Compiling RetroClash.Stack ( src/RetroClash/Stack.hs, dist/build/RetroClash/Stack.o, dist/build/RetroClash/Stack.dyn_o )
[ 6 of 19] Compiling RetroClash.Utils ( src/RetroClash/Utils.hs, dist/build/RetroClash/Utils.o, dist/build/RetroClash/Utils.dyn_o )
[ 7 of 19] Compiling RetroClash.SevenSegment ( src/RetroClash/SevenSegment.hs, dist/build/RetroClash/SevenSegment.o, dist/build/RetroClash/SevenSegment.dyn_o )
[ 8 of 19] Compiling RetroClash.SerialTx ( src/RetroClash/SerialTx.hs, dist/build/RetroClash/SerialTx.o, dist/build/RetroClash/SerialTx.dyn_o )
[ 9 of 19] Compiling RetroClash.SerialRx ( src/RetroClash/SerialRx.hs, dist/build/RetroClash/SerialRx.o, dist/build/RetroClash/SerialRx.dyn_o )
[10 of 19] Compiling RetroClash.PS2   ( src/RetroClash/PS2.hs, dist/build/RetroClash/PS2.o, dist/build/RetroClash/PS2.dyn_o )
[11 of 19] Compiling RetroClash.PS2.ASCII ( src/RetroClash/PS2/ASCII.hs, dist/build/RetroClash/PS2/ASCII.o, dist/build/RetroClash/PS2/ASCII.dyn_o )
[12 of 19] Compiling RetroClash.Memory ( src/RetroClash/Memory.hs, dist/build/RetroClash/Memory.o, dist/build/RetroClash/Memory.dyn_o )

src/RetroClash/Memory.hs:64:32: error:
    • No instance for (Monoid (Q [Dec]))
        arising from the 'deriving' clause of a data type declaration
      Possible fix:
        use a standalone 'deriving instance' declaration,
          so you can specify the instance context yourself
    • When deriving the instance for (Applicative (Addressing addr))
   |
64 |     deriving newtype (Functor, Applicative, Monad)
   |                                ^^^^^^^^^^^

src/RetroClash/Memory.hs:64:45: error:
    • No instance for (Monoid (Q [Dec]))
        arising from the 'deriving' clause of a data type declaration
      Possible fix:
        use a standalone 'deriving instance' declaration,
          so you can specify the instance context yourself
    • When deriving the instance for (Monad (Addressing addr))
   |
64 |     deriving newtype (Functor, Applicative, Monad)
   |                                             ^^^^^
[13 of 19] Compiling RetroClash.Keypad ( src/RetroClash/Keypad.hs, dist/build/RetroClash/Keypad.o, dist/build/RetroClash/Keypad.dyn_o )
[14 of 19] Compiling RetroClash.I2C   ( src/RetroClash/I2C.hs, dist/build/RetroClash/I2C.o, dist/build/RetroClash/I2C.dyn_o )
[15 of 19] Compiling RetroClash.CPU   ( src/RetroClash/CPU.hs, dist/build/RetroClash/CPU.o, dist/build/RetroClash/CPU.dyn_o )
[16 of 19] Compiling RetroClash.BCD   ( src/RetroClash/BCD.hs, dist/build/RetroClash/BCD.o, dist/build/RetroClash/BCD.dyn_o )
[17 of 19] Compiling RetroClash.VGA   ( src/RetroClash/VGA.hs, dist/build/RetroClash/VGA.o, dist/build/RetroClash/VGA.dyn_o )
[18 of 19] Compiling RetroClash.Delayed ( src/RetroClash/Delayed.hs, dist/build/RetroClash/Delayed.o, dist/build/RetroClash/Delayed.dyn_o )
[19 of 19] Compiling RetroClash.Video ( src/RetroClash/Video.hs, dist/build/RetroClash/Video.o, dist/build/RetroClash/Video.dyn_o )
error: builder for '/nix/store/xgakbif7d20j1rjar925r426pr4br37x-retroclash-lib-0.1.2.drv' failed with exit code 1
error: 1 dependencies of derivation '/nix/store/72zmq94vlkasy9w1k7mnvbnhvvfm7gcl-ghc-8.10.7-with-packages.drv' failed to build
```

**Pull request merged**